File: Blind150/1.Array&Hash/GO/equal_string_one_swap.go
// https://leetcode.com/problems/check-if-one-string-swap-can-make-strings-equal/description/
func canBeSwapped(s1, s2 string) bool {
	if len(s1) != len(s2) {
		return false
	}

	if s1 == s2 {
		return true
	}

	setS1 := make(map[byte]struct{})
	setS2 := make(map[byte]struct{})
	notMatched := 0

	for i := 0; i < len(s1); i++ {
		x := s1[i]
		y := s2[i]

		if x != y {
			setS1[x] = struct{}{}
			setS2[y] = struct{}{}

			notMatched++

			if notMatched > 2 {
				return false
			}
		}
	}

	if len(setS1) != len(setS2) {
		return false
	}

	for key := range setS1 {
		if _, exists := setS2[key]; !exists {
			return false
		}
	}

	return true
}

File: Blind150/1.Array&Hash/Python/equal_string_one_swap.py
# https://leetcode.com/problems/check-if-one-string-swap-can-make-strings-equal/description/
class Solution:
    def areAlmostEqual(self, s1: str, s2: str) -> bool:

        if len(s1) != len(s2):
            return False

        if s1 == s2:
            return True

        set_s1 = set()
        set_s2 = set()
        not_matched = 0


        for x,y in zip(s1,s2):

            if x != y:
                set_s1.add(x)
                set_s2.add(y)

                not_matched += 1

                if not_matched > 2:
                    return False


        return True if set_s1 == set_s2 else False
